Blog post: No-code experiments are a lie (#15429)
PostHog/posthog.com: contents/blog/no-code-experiments-are-a-lie.md

Reviewer: We need to update the file name to be `ai-is-killing-no-code-experiments`

---
title: "No-code experiments are a lie"
date: 2026-03-03
author: ["leonhard-prinz"]
tags:
  - Experiments
  - Growth engineering
---

Every no-code A/B testing tool makes the same pitch: growth runs experiments, engineers ship product, nobody waits on anybody.

> The key problem with this approach though is that these tools inject changes _after_ the page renders. Your browser loads the original page, then the tool frantically swaps in the variant. Users see a flash of the original content before the change kicks in. That's flicker, and it is just not a good experience.

Reviewer 2: > these tools inject changes _after_ the page renders

Not necessarily. You can run a no-code experiments on a server-rendered application, and it should work correctly and without a flicker.

The problem is when using them:
a) On a webpage after it's been loaded in the frontend - causes a flicker
b) On a single page application which frequently re-renders and thus the DOM changes are reverted.

We document these limitations in the docs: https://posthog.com/docs/experiments/no-code-web-experiments#limitations-and-use-cases

---
title: "No-code experiments are a lie"
date: 2026-03-03
author: ["leonhard-prinz"]
tags:
  - Experiments
  - Growth engineering
---

> Every no-code A/B testing tool makes the same pitch: growth runs experiments, engineers ship product, nobody waits on anybody.

Reviewer: ```suggestion
Every no-code A/B testing tool makes the same pitch: run experiments without relying on engineers.
```

2. Feeds it to AI, either through [PostHog AI](link) in-app (alpha, requires GitHub connection) or locally in your IDE via the [MCP connector](link) (available now)
3. Reviews the generated code + experiment config
4. Opens the PR
5. Monitors results in PostHog
6. Decides: roll out the winner, iterate, or kill it

**Engineering team:**

1. Reviews PRs for code quality and unintended side effects
2. Merges approved PRs
3. Handles cleanup when experiments conclude

Growth moves at their own pace instead of camping in the engineering backlog. Engineering maintains code quality without being the bottleneck. It removes a lot of the moving parts and dependencies and swaps it for speed and ownership.

| | Traditional no-code | AI code generation |
|---|---|---|
| Growth creates experiment | Visual editor | AI prompt |
| Dev builds experiment | Not needed (upfront) | Not needed (AI generates) |
| Dev reviews | Not needed (upfront) | Reviews a PR. Minutes for simple changes |
| Experiment runs | With flicker risk on SPAs | No injection issues. No flicker with server-side flag eval |
| Winner found | Need to rebuild in real code | Flip feature flag to 100%. Done |

> | Cleanup | Rebuild winning variant + remove test from tools | Remove flag wrapper + dead variant code (soon automated) |

Reviewer: I don't fully understand this table, it definitely needs a "lead in" explaining what it is.

The left side is the steps? Maybe something like this would work?

```suggestion
Growth moves at their own pace instead of camping in the engineering backlog. Engineering maintains code quality without being the bottleneck. It removes a lot of the moving parts and dependencies and swaps it for speed and ownership. The process changes to look like this:

| | Traditional no-code | AI code generation |
|---|---|---|
| 1. Growth creates experiment | Visual editor | AI prompt |
| 2. Dev builds experiment | Not needed (upfront) | Not needed (AI generates) |
| 3. Dev reviews | Not needed (upfront) | Reviews a PR. Minutes for simple changes |
| 4. Experiment runs | With flicker risk on SPAs | No injection issues. No flicker with server-side flag eval |
| 5. Winner found | Need to rebuild in real code | Flip feature flag to 100%. Done |
| 6. Cleanup | Rebuild winning variant + remove test from tools | Remove flag wrapper + dead variant code (soon automated) |
```

Content and copy changes are the fastest wins, but because this is real code (not just HTML and CSS overlays), you can go beyond surface-level experiments. Layouts, multi-step forms, conditional UI based on user segments. Anything you can describe, AI can generate. For more complex changes, be even more specific about spacing, element order, and responsive behavior.

This brief translates directly into a prompt. Open your editor with the [PostHog MCP server](/docs/model-context-protocol) connected and start the conversation:

> "Create an A/B test on the signup page. The control keeps the current 'Get started' CTA. The variant should change it to 'Start your free 14-day trial' and add 'No credit card required' below the button. Split traffic 50/50. Track signup button click-through rate as the primary metric."

You don't need to get everything right in one message. PostHog AI will look at your code, ask about edge cases, and suggest things you might have missed. After a few back-and-forth messages, it generates the code change, the experiment in PostHog, and all the configuration. Review what it produced, then open the PR.

## Previewing before you launch

> The experiment deploys behind a feature flag at 0% rollout. No users see it. From there, three ways to preview:

Reviewer: ```suggestion
Many rely on the visual preview no-code experiments provide and might be worried about losing them, but again thanks to agents, this is solvable.

The experiment deploys behind a feature flag at 0% rollout. No users see it. From there, three ways to preview:
```

> 1. **Run locally.** Pull the branch, run your dev server, and toggle the flag in your local environment. You see the variant with real styling and real components.

Reviewer: ```suggestion
1. **Run locally.** Pull the branch, run your dev server, and toggle the flag in your local environment. You see the variant with real styling and real components. If you get stuck, use an agent to help you debug.
```
